server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,260 @@
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Helper controller class to remove list of nodes from the cluster and update status
+ *
+ * @opensearch.internal
+ */
+
+public class DecommissionController {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionController.class);
+
+    private final NodeRemovalClusterStateTaskExecutor nodeRemovalExecutor;
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final ThreadPool threadPool;
+
+    DecommissionController(
+        ClusterService clusterService,
+        TransportService transportService,
+        AllocationService allocationService,
+        ThreadPool threadPool
+    ) {
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.nodeRemovalExecutor = new NodeRemovalClusterStateTaskExecutor(allocationService, logger);
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Transport call to add nodes to voting config exclusion
+     *
+     * @param nodes set of nodes Ids to be added to voting config exclusion list
+     * @param listener callback for response or failure
+     */
+    public void excludeDecommissionedNodesFromVotingConfig(Set<String> nodes, ActionListener<Void> listener) {
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            AddVotingConfigExclusionsAction.NAME,
+            new AddVotingConfigExclusionsRequest(
+                Strings.EMPTY_ARRAY,
+                nodes.toArray(String[]::new),
+                Strings.EMPTY_ARRAY,
+                TimeValue.timeValueSeconds(30)
+            ),
+            new TransportResponseHandler<AddVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(AddVotingConfigExclusionsResponse response) {
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    listener.onFailure(exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public AddVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new AddVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * Transport call to clear voting config exclusion
+     *
+     * @param listener callback for response or failure
+     */
+    public void clearVotingConfigExclusion(ActionListener<Void> listener) {
```

Comment:
Yeah I agree selective clearing of exclusion could lead to code complication and introducing new bugs.

I can think of 2 options:
1. We take input in the decommission api to clear the voting configuration and tell that it will clear existing exclusions as well. So that user sets that option consciously.
2. Or we always leave it as is and user can remove them later own their own.

Reply:
> Or we always leave it as is and user can remove them later own their own

This can have different consequences I think. For eg. It was a 3 zone cluster with one leader in each. We went ahead and decommissioned one zone. Due to which the current state has one exclusion and only one leader in VotingConfiguration (from what I understand the config will always have odd number).
Now, the user has recommissioned the zone back but since we didn't wipe off the exclusion, the leader from zone which was decommissioned is still part of exclusion. And because of this VotingConfig size still remains 1 (which is unexpected after recommissioning)
Basically, after recommissioning, we should not leave any change that was done due to decommission as is. 

> We take input in the decommission api to clear the voting configuration and tell that it will clear existing exclusions as well. So that user sets that option consciously.

This looks a better option to me. But what if, the user says no don't wipe off my exclusions and in that case, we cannot even decommission. And hence, this flag in a way mandatory for decommission. Now since it is mandatory, we cannot really call it a flag.

Reply:
It would have made sense if we could have guaranteed that we don't clear voting configuration at all or clear if no previous exclusion set. However since there is no guarantee that no voting config exclusion changes would have been made since the decommission started, it makes decommission API *always* have to clear and possibly also override user exclusions. Hence documenting this behaviour for exclusion seems fine